lib/src/kbasesearchengine/main/IndexerWorker.java

```diff
@@ -844,18 +808,139 @@ private void indexObjectWrapperFn(final List<Object> input)
+                resRef = new ResolvedReference(resRef.getReference(),
+                                               guid,
+                                               resRef.getType(),
+                                               resRef.getTimestamp());
+
+                // get object from data source
+                final File tempFile;
+                final File subDir;
+                try {
+                    subDir = FileUtil.getOrCreateSubDir(rootTempDir, guid.getStorageCode());
+                    tempFile = File.createTempFile("ws_srv_response_", ".json", subDir);
+                } catch (IOException ex) {
+                    throw new RetriableIndexingException(ErrorType.OTHER, ex.getMessage(), ex);
+                }
+
+                try {
+                    final EventHandler handler = getEventHandler(guid);
+                    List<GUID> refPath = new ArrayList<>(objectRefPath);
+                    refPath.add(resRef.getReference());
+                    final SourceData obj = handler.load(refPath, tempFile.toPath());
+
+                    final ObjectTypeParsingRules rules;
+                    try {
+                        rules = getObjectTypeParsingRules(resRef);
+                    } catch (ObjectParseException ex ) {
+                        throw new UnprocessableEventIndexingException(ErrorType.OTHER,
+                                ex.getMessage());
+                    }
+
+                    final ParseObjectsRet parsedRet = parseObjects(guid, this,
+                            Arrays.asList(resRef.getResolvedReference()), obj, rules);
+
+                    final GUID parsedGUID = parsedRet.guidToObj.keySet().iterator().next();
+
+                    final ObjectData objData = buildObjectDataFrom(
+                            guid,
+                            rules.getGlobalObjectType(),
+                            obj.getName(),
+                            obj.getCreator(),
+                            obj.getCommitHash(),
+                            parsedRet.guidToObj.get(parsedGUID).getKeywords());
+
+                    data.add(objData);
+                } finally {
+                    tempFile.delete();
+                }
             }
+            return data;
+        }
+
+        private ObjectTypeParsingRules getObjectTypeParsingRules(
+                ResolvedReference ref) throws ObjectParseException {
+
+            final List<ObjectTypeParsingRules> parsingRules = new ArrayList<>(
+                    typeStorage.listObjectTypeParsingRules(ref.getType()));
+
+            // filter by subObject type
+            final String refSubObjectType = ref.getResolvedReference().getSubObjectType();
+            for (final ObjectTypeParsingRules rule : parsingRules) {
+                final String ruleSubObjectType = rule.getSubObjectType().orNull();
+                if (refSubObjectType == null && ruleSubObjectType == null) {
```

Comment:
I think this is the case because sub-object types key off the parent type. So even if two parents had identical sub-objects, the sub-obj types would have to be defined separately with each one keying off of its respective parent.

Reply:
sub-object type specs i.e.

Reply:
> I think this is the case because sub-object types key off the parent type. 

Is that true? I'm looking at the AssemblyContig spec and it doesn't look to me like there's any references to the Assembly type at all. IFAIK the Assembly spec could be deleted and everything would work fine. Am I incorrect here?

Reply:
I guess keying off is not the right terminology. The line below states that AssemblyContig is an inner-sub-type of the parent type Assembly right? This is what I meant.   

storage-object-type: KBaseGenomeAnnotations.Assembly

Reply:
Both the search types Assembly and AssemblyContig relate to the same storage type (e.g. workspace type), KBaseGenomeAnnotations.Assembly, but AFAICT are otherwise completely independent. You could have search installations that index both, only Assembly search types, or only AssemblyContig search types. AFAIK there could also be other non-subobject types related to the source type.

To the best of my ability to tell, that means that when the code assumes there's always one top-level / non-subobject type per source type, as I think it does in this case, that's a bug.

Reply:
By source type, do you mean storage type?

Reply:
Yes, edited to make the names consistent.

Reply:
I think I see what you mean now. I was thinking in a different direction (a sub-object having two parent types). But I think what you mean is that there can be many specs defined for a single storage object type. So in this case, we would need to populate more than one index for each guid with a storage object type that maps to multiple search object types (sub-object types seem to add a layer of complexity and will also have to be handled in a similar manner). Does this rephrasing of the problem/solution sound right? I want to make sure before I make the change since the change cascades into other parts.

Reply:
> (a sub-object having two parent types)

I don't think I understand what you mean by this.

Your rephrasing sounds right, but just in case, here's another rephrasing:

For each storage type, there may be X search type specs and Y search sub-type specs, where X and Y are integers from 0 to N. There is no relationship between the various specs other than they happen to be for the same storage type.

The current code handles this situation correctly because it just pulls the list of search types for the storage type from the type storage system and processes them one by one. This new code, if I understand it correctly, does not do that and makes assumptions about the values of X and Y.

Reply:
After analyzing it more, I think the app is still working as we expect it to. The reason for this is that all X types refer back to the same storage type. The same is true for the Y sub-types. So the code in the snippet above will get back that storage type. 

As a sanity check I ran this PR version against CI with two genome.yaml files (with search types Genome and GenomeNew) referring back to the same storage type and added the mapping shown below and it generated the corresponding genome_2 and genomeNew_2 indexes. This I think is also the use case that Shane described in our reindexing discussion yesterday. 

```
storage-type: WS
types:
    KBaseGenomes.Genome:
        default:
           -
               type: Genome
               version: 2
           -
               type: GenomeFeature
               version: 2
           -
               type: GenomeNew
               version: 2
```

Reply:
I don't think it can be working right. I think the bug will occur if 1) there's 2 or more specifications for a storage type and 2) a recursion encounters that type.

What the code is doing is, given a storage type and a guid (which in the case of the bug, does not have a sub-object type):

1) Get all the object type parsing rules for the storage type, of which there are N
2) Recurses though the rules until it finds one without a subobject type
3) Returns that rule.

That means that any other rules without sub-object types are ignored.

Actually, on reading further, it's the same problem for sub-object types - it finds the first rule that matches the sub-object type  and returns it, ignoring any other rules.

Reply:
The reason why it looks like it is working correctly is because the objects get indexed in the code below. 

https://github.com/kbase/KBaseSearchEngine/blob/develop/lib/src/kbasesearchengine/main/IndexerWorker.java#L538

The code snippet above is used in two places.

The code is being used to perform a lookup transform here -

https://github.com/kbase/KBaseSearchEngine/blob/develop/lib/src/kbasesearchengine/parse/KeywordParser.java#L313

The code is being used to find the search object type here - 

https://github.com/kbase/KBaseSearchEngine/blob/develop/lib/src/kbasesearchengine/parse/KeywordParser.java#L294

In both cases, the code is being used for performing transforms. This code will blow up only when a transform with a reference to a storage type that maps to two or more search types is made. This code AFAICT is not a regression and points to inherent bugs in the application. I am now working on fixing this.